src/Components/Web.JS/src/StreamingInterop.ts

```diff
@@ -0,0 +1,54 @@
+export async function getNextChunk(data: ArrayBufferView | Blob, position: number, nextChunkSize: number): Promise<Uint8Array> {
+    if (data instanceof Blob) {
+        return await getChunkFromBlob(data, position, nextChunkSize);
+    } else {
+        return getChunkFromArrayBufferView(data, position, nextChunkSize);
+    }
+}
+
+async function getChunkFromBlob(data: Blob, position: number, nextChunkSize: number): Promise<Uint8Array> {
+    const chunkBlob = data.slice(position, position + nextChunkSize);
+    const arrayBuffer = await chunkBlob.arrayBuffer();
+    const nextChunkData = new Uint8Array(arrayBuffer);
+    return nextChunkData;
+}
+
+function getChunkFromArrayBufferView(data: ArrayBufferView, position: number, nextChunkSize: number) {
+    const nextChunkData = new Uint8Array(data.buffer, data.byteOffset + position, nextChunkSize);
+    return nextChunkData;
+}
+
+export async function sendJSDataStreamUsingObjectReference(
+        send: (streamId: number, chunkId: number, nextChunkData: Uint8Array | null, error: string | null) => Promise<boolean>,
+        data: ArrayBufferView | Blob,
+        streamId: number,
+        chunkSize: number) {
+
+    // Run the rest in the background, without delaying the completion of the call to sendJSDataStreamUsingObjectReference
+    // otherwise we'll deadlock (.NET can't begin reading until this completes, but it won't complete
+    // because nobody's reading the pipe)
+    setTimeout(async () => {
+        try {
+            const byteLength = data instanceof Blob ? data.size : data.byteLength;
+            let position = 0;
+            let chunkId = 0;
+
+            while (position < byteLength) {
+                const nextChunkSize = Math.min(chunkSize, byteLength - position);
+                let nextChunkData = await getNextChunk(data, position, nextChunkSize);
+
+                const streamIsAlive = await send(streamId, chunkId, nextChunkData, null);
```

Comment:
At first I was surprised to see a push-based transfer mechanism for WebAssembly, because I would have expected this to completely block the UI thread until the transfer is complete. Does it not block the UI thread because, on the .NET side, it's feeding things into a `Pipe` which imposes backpressure?

TBH I would have expected a much simpler model to be pull-based for WebAssembly and WebView. The actual `Stream` instance could directly fetch an arbitrarily large block of data in a single chunk, whenever the .NET side tried to read from the stream. There would be no `Pipe` and much of the other orchestration logic would disappear completely. We wouldn't need to do anything tricky with layering to share logic between Server and WebAssembly/WebView, because the WebAssembly/WebView code would be very minimal and different from Server (though identical for the two of them).

Is that not possible, or do you think it's actually more complicated? Or is it worth trying?

Reply:
Also more comments about this at https://github.com/dotnet/aspnetcore/pull/33986/files#r668052504

Reply:
Thanks Steve for the review and comments, I've updated the implementation accordingly to utilize a pull based approach in which we pull from JS when the `JSDataStream.ReadAsync` is called.